cmd/kube-addon-manager/app/addon-manager.go

```diff
@@ -0,0 +1,349 @@
+func appendPruneWhitelistFlags(pwf, resources []string) []string {
+	for _, resource := range resources {
+		pwf = append(pwf, "--prune-whitelist", resource)
+	}
+	return pwf
+}
+
+func (m *addonManager) Run() error {
+	m.waitForSystemServiceAccount()
+	if err := m.ensureDefaultAdmissionControlsObjects(); err != nil {
+		if os.IsNotExist(err) {
+			klog.Warningf("admissions control object directory not found: %v", err)
+		} else {
+			klog.Errorf("could not create admissions control objects: %v", err)
+		}
+	}
+	for {
+		start := time.Now()
+
+		if !m.leaderElection || m.isLeader() {
+			m.ensureAddons()
+			m.reconcileAddons()
+		} else {
+			klog.Info("== Another addon-manager is acting as leader. Sleeping... ==")
+		}
+
+		elapsed := time.Since(start)
+		if elapsed < m.checkInterval {
+			time.Sleep(m.checkInterval - elapsed)
+		}
+	}
+}
+
+func (m *addonManager) waitForSystemServiceAccount() {
+	var out, errOut bytes.Buffer
+	for {
+		err := m.kubectl(&out, &errOut, m.kubectlOpts, "get", "-n", systemNamespace, "serviceaccount", "default", "-o", "go-template={{with index .secrets 0}}{{.name}}{{end}}")
+		if errStr := errOut.String(); errStr != "" {
+			klog.Warning(errStr)
+		}
+		if err == nil && out.String() != "" {
+			return
+		}
+		out.Reset()
+		errOut.Reset()
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// Create admission_control objects if defined before any other addon services. If the limits
+// are defined in a namespace other than default, we should still create the limits for the
+// default namespace.
+func (m *addonManager) ensureDefaultAdmissionControlsObjects() error {
+	var out, errOut bytes.Buffer
+	return filepath.Walk(m.admissionControlsDir,
+		func(path string, _ os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			ext := filepath.Ext(path)
+			if ext == ".yaml" || ext == ".json" {
+				for attemptsRemaining := startAddonAttempts; attemptsRemaining > 0; attemptsRemaining-- {
+					err = m.kubectl(&out, &errOut, m.kubectlOpts, "--namespace=default", "apply", "-f", path)
+					filterLog(&out, "", klog.Info)
+					filterLog(&errOut, "", klog.Warning)
+					if err == nil {
+						klog.Infof("== Successfully started %s in namespace default ==", path)
+						return nil
+					}
+					klog.Warningf("== Failed to start %s in namespace default. %d tries remaining. ==", path, attemptsRemaining-1)
+					if attemptsRemaining > 1 {
+						time.Sleep(startAddonRetryDelay)
+					}
+				}
+			}
+			return err
+		})
+}
+
+// IsLeader returns whether this instance of addonManager should act as leader in a multi-master cluster.
+func (m *addonManager) isLeader() bool {
+	var out, errOut bytes.Buffer
+	klog.Info("== Determining addon-manager leader ==")
+	err := m.kubectl(&out, &errOut, m.kubectlOpts, "-n", systemNamespace, "get", "ep", "kube-controller-manager", "-o", `go-template={{index .metadata.annotations "control-plane.alpha.kubernetes.io/leader"}}`)
```

Comment:
This is just gross. I realize this is legacy code and we need to make sure this code works correctly during an upgrade. (So we need to think through how to safely move off of the KCM lock) However piggy backing on the KCM's leader is *NOT* a good idea. 1) There are 3 possible resources for leader election, so we are not assured that its endpoints. 2) I believe the kube-up default is migrating away from endpoints and this will be broken when that migration completes. 3) The KCM leader is currently our heaviest burden on a control plane node. Deliberately co locating other process leaders with the KCM leader exacerbates this issue. 4) Leader election is timing sensitive with it durations. This code is beginning its belief it holds the lock value at some point after the lock was renewed and pays no attention to when the lock expires. As such it is not providing as good a guarantee to the addon manager as it does to the KCM.

Can we file as issue and provide a TODO in the code here related to the issue?